jetty-ee10/jetty-ee10-osgi/jetty-ee10-osgi-boot/src/main/java/org/eclipse/jetty/ee10/osgi/annotations/AnnotationParser.java

```diff
@@ -109,25 +109,28 @@ public void parse(final Set<? extends Handler> handlers, Resource r) throws Exce
         if (!r.exists())
             return;
 
-        if (FileID.isJavaArchive(r.getPath()))
-        {
-            parseJar(handlers, r);
-            return;
-        }
-
         if (r.isDirectory())
         {
             parseDir(handlers, r);
             return;
         }
-
-        if (FileID.isClassFile(r.getPath()))
+        else
         {
-            parseClass(handlers, null, r.getPath());
+            if (FileID.isJavaArchive(r.getFileName()))
```

Comment:
Why the change in order?

Reply:
To make the conditions easier.

Keeping the order it would have been ...

``` java
if (!r.isDirectory() && FileID.isJavaArchive(r.getFileName()))
        {
            parseJar(handlers, r);
            return;
        }

        if (r.isDirectory())
        {
            parseDir(handlers, r);
            return;
        }

        if (!r.isDirectory() && FileID.isClassFile(r.getFileName()))
        {
            parseClass(handlers, null, r.getPath());
        }
```

Which reads really weird.
So, it was changed to be anything directory related do this, anything else (a file), do that.

Reply:
Well, the point of this PR is just to add null protection, so I don't see how a change in the order of the way we consider what to do with a given resource is related. I don't like changing the well-established order of things unless there is an associated bug.

Reply:
The NPE protection is just part of the change.
The change also moves away from Path to Resource, so that the example of URLResource has an opportunity to work.
That change means we also change this section.
The new parseClass() that uses Resource is another example of this change.
Let me change the title of this PR, the description still has the details.